### ISSUE TITLE
Login failed

### ISSUE THREAD

**Issue by Pooort** [tedchou12/webull#187]
Login failed with:
```
id = str(result['data'][0]['secAccountId'])
IndexError: list index out of range
```

**Comment by PageMastr** [comment]
code snippet?

comment out that line and add a print(str(result)) and see whats coming back.

Mostly likely too many failed validations

**Comment by kucharskim** [comment]
Got same issue. Was just playing with `webull` Python code. My account is not fully registered, as it turned out Webull is not available in my country. Sample code:

```
from webull import webull

wb = webull()
a = wb.login('email@address.example.com', 'password')
b = wb.get_account_id()
c = wb.get_account()

print(a)
print(b)
print(c)
```

Crash:

```
(webull-env) $ python we.py
Traceback (most recent call last):
  File "we.py", line 4, in <module>
    a = wb.login('email@address.example.com', 'password')
  File "/virtualenv3/webull-env/lib/python3.8/site-packages/webull/webull.py", line 140, in login
    self._account_id = self.get_account_id()
  File "/virtualenv3/webull-env/lib/python3.8/site-packages/webull/webull.py", line 274, in get_account_id
    self.zone_var = str(result['data'][0]['rzone'])
IndexError: list index out of range
```

More info:

```
--- webull/webull.py
+++ webull/webull.py
@@ -271,6 +271,9 @@ class webull:
         response = requests.get(self._urls.account_id(), headers=headers)
         result = response.json()
         if result['success'] :
+            from pprint import pprint
+            import sys
+            pprint(result, stream=sys.stderr)
             self.zone_var = str(result['data'][0]['rzone'])
             id = str(result['data'][0]['secAccountId'])
             self._account_id = id
```

Results:

```
(webull-env) $ python we.py                                                                                           
{'data': [], 'success': True}
Traceback (most recent call last):
  File "we.py", line 4, in <module>
...
```

**Comment by kucharskim** [comment]
I wrote above, got the same issue. It's not the same, but similar, so actually my commit may not address problem reported here originally.

**Comment by PageMastr** [comment]
Please go and read the wiki. You are not logging in correctly in the code
snippet you provided.  MFA and Security question responses are MANDATORY in
order to login successfully.

Until you have logged in successfully, you will continue to get this error.

**Comment by PageMastr** [comment]
https://github.com/tedchou12/webull/wiki
https://github.com/tedchou12/webull/wiki/MFA-&-Security

**Comment by kucharskim** [comment]
Hm.. then could the message be more clear? Two people made a mistake in a window of 2 weeks. If you leave error as is:

      IndexError: list index out of range

You will get more reports like this.

**Comment by PageMastr** [comment]
Secondary Ticket Close reason:
Not enough information provided to investigate possible bug.
Request for more information has resulted in no reply from OP

**Comment by PageMastr** [comment]
> Hm.. then could the message be more clear? Two people made a mistake in a window of 2 weeks. If you leave error as is:
> 
> ```
>   IndexError: list index out of range
> ```
> 
> You will get more reports like this.

I have updated the Wiki, so that the examples now include updated mfa/security code.

This issue is only possible in this context when a failed login occurs.

**Comment by kucharskim** [comment]
Error was not failed login, error was `IndexError: list index out of range`.

**Comment by kucharskim** [comment]
I did more testing. Login was successful. I believe my diff was correct.

I've specified wrong username and password, result:

```
login()
{'code': 'phone.illegal',
 'msg': 'Wrong format for mobile phone number',
 'traceId': '<random id>'}
```

Correct username and password:

```
login()
{'accessToken': '<random string>',
 'extInfo': {'userPwdFlag': '1'},
 'firstTimeOfThird': False,
 'refreshToken': '<random string>',
 'registerAddress': 137,
 'settings': {'chartOption': 2,
              'createTime': '2021-03-07T16:06:59.000+0000',
              'focusMarketId': '2,3,4,5,6,14',
              'fontSize': 'M',
              'hotNews': 1,
              'id': <id number>,
              'increDecreColor': 2,
              'kdata': 1,
              'language': 'en',
              'languageUpdateTime': '1970-01-01T00:00:00.000+0000',
              'listStyle': 1,
              'operateTime': '1969-12-31T00:00:00.000+0000',
              'orderDealRemind': 1,
              'portfolioBulletin': 1,
              'portfolioDisplayMode': 2,
              'portfolioHoldingsDisplay': 1,
              'portfolioIndexDisplay': 1,
              'portfolioNameNewline': 1,
              'refreshFrequency': 1,
              'regionId': 137,
              'shock': 0,
              'theme': 2,
              'tickerPriceRemind': 1,
              'updateTime': '2021-03-07T16:06:59.000+0000',
              'userId': <user id>},
 'tokenExpireTime': '2021-05-06T22:16:18.637+0000',
 'uuid': '<random string>'}
```

**Comment by PageMastr** [comment]
> Error _message_ was not failed login, error was `IndexError: list index out of range`.

Here is exactly why what you want, cannot be done.

This Error Message, ONLY happens, if a login was not successful. 

Login code is thoroughly documented with multiple available examples on the wiki.

Had the login portion had debug code, you would have gotten the error response for not including MFA and security question info.

And your not refreshing a token in that snippet anywhere using wb.refresh_login() so Theres no existing Auth Token information.

And lastly, Your response, is not formatted correctly above. as that is not a valid return json response from the webull api. Its out of order. 

Here is an example of a succesfull Login response. with my information blocked out. 

![avav345](https://user-images.githubusercontent.com/9737536/110257115-f8f08680-7f61-11eb-86d0-a3d2ddb76675.PNG)

**Comment by kucharskim** [comment]
My last examples are with my diff included otherwise it would crash with correct username and password.

Regards your last comment with screenshot, did you read my comments?

**Comment by PageMastr** [comment]
> My last examples are with my diff included otherwise it would crash with correct username and password.
> 
> Regards your last comment with screenshot, did you read my comments?

Yes I did. 

again, your not logged in.

As you stated above, Webull is not available in your country, so your account is not active, ergo you cannot fully login and fetch account id details etc. 

Which circles back to, Its a login issue, people with the issue should debug their login code. 
Its documented in the wiki. 

Better Error logging is not needed, for a situation, in which the apparent only other way to get this error. Is by opening an account in a country where webull is not available.

**Comment by kucharskim** [comment]
> Better Error logging is not needed, for a situation, in which the apparent only other way to get this error

Then I think lack of account ID should be handled better in the code, and clear exception should be returned to the user. If account id is `None` throw exception that this is a problem, so user can focus their troubleshooting efforts on the account setup side and not on the Python code side. User experience is lacking here.

**Comment by PageMastr** [comment]
Addressing first error message for incorrect login, Your debug code catches this. You didnt login.  so you didnt get the issue fetching account info.

On second one, With correct account info. "My account is not fully registered, as it turned out Webull is not available in my country"

This is why you get the error.  and adding error logging to tell you your account might not be valid is far outside of the scope of an API. 

Normal users, who have accounts that are fully registered and active, will not receive this error on successful login.

**Comment by kucharskim** [comment]
I'm so confused the way your drive this discussion. I don't know are we on the same page.

> This is why you get the error

What error? Can you be specific? I got only one error which I'm ware of `IndexError: list index out of range`. Are you referring to that `IndexError`?

**Comment by PageMastr** [comment]
TLDR, it is unreasonable to demand, error logging for "IndexError: list index out of range" , that is the result of:
1.) an account not valid for using webull - as this is an API for using webull. something your account is not available for.
2.) Failure to properly debug own code to verify successful login.

**Comment by kucharskim** [comment]
> I think ive figured out where the communication issue is.

Thank you, as during this whole discussion I had a feeling we are no on the same page. This paints better picture for me.

**Comment by PageMastr** [comment]
> > I think ive figured out where the communication issue is.
> 
> Thank you, as during this whole discussion I had a feeling we are no on the same page. This paints better picture for me.

anytime, and its ok, I was feeling the same way haha. Just needed a bit more detail to it on my part. Ill try to remember that for the future haha.

**Comment by kucharskim** [comment]
Probably closing comment from me..

> Failure to properly debug own code to verify successful login.

I don't agree. User needs to debug your code to figure this out.

> an account not valid for using webull - as this is an API for using webull. something your account is not available for

Right, but then better user experience could be achieved here. However, as you wrote earlier:

> adding error logging to tell you your account might not be valid is far outside of the scope of an unofficial API.

Sums up, that my request of better user experience is not within of your scope. I get it. I don't think here is more to discuss. We just have different opinions about the approach.

I still think my diff was correct or code should throw more meaningful exception.

Problem for me is, with my diff I can use your code. I can use `active_gainer_loser()`, `get_analysis()`, `get_tradable()`, etc. Buy maybe those functions are available without `login()`, so maybe I don't need account with Webull to use module.

**Comment by PageMastr** [comment]
> > Failure to properly debug own code to verify successful login.
> 
> I don't agree. User needs to debug your code to figure this out.
> 
The reason i disagree with this, is because its documented how to login, and that you need More than just email and password.
had the login response been debugged once to check if login was successful (in the case of an active account) they would have not made it to this error.

> > adding error logging to tell you your account might not be valid is far outside of the scope of an unofficial API.
> 
> Sums up, that my request of better user experience is not within of your scope. I get it. I don't think here is more to discuss. We just have different opinions about the approach.
> 
I should rephrase this, its not within the scope at this time. The reason is because this is an unofficial api. I was only able to recently fix logging in, which was broken for a while due to the V5 api upgrade by webull restricting new accounts from logging in. They change endpoints and update the api constantly, and at this point, we dont even have all of the functions available, into the api. For instance, you can call account graph data from webull, but that functionality is not possible yet through this api.

Improving user experience is a long ways off, if even potentially possible, and will be entirely determined by community help, someone taking the time or having time to implement it etc. and those of us who work on this regularly and are committed to this project, are mostly focused on fixing major bugs, and mapping and adding functions for unexplored parts of the API.


> I still think my diff was correct or code should throw more meaningful exception.
> 
I agree, its definitely something, that can be done. And i would invite you, to make those changes and submit a pull request so that they can be added. Its just something that for the reasons above, isnt something that is likely to be worked on, or have time put into, for a significant amount of time, just due to the amount of time, it takes just to keep the api working.

> Problem for me is, with my diff I can use your code. I can use `active_gainer_loser()`, `get_analysis()`, `get_tradable()`, etc. Buy maybe those functions are available without `login()`, so maybe I don't need account with Webull to use module.
This is correct, those particular functions are available without logging in.

**Comment by kucharskim** [comment]
I did have `accessToken` and `refreshToken` in response received with my diff included, so I believe `login()` was successful.

Do you think if I will go through the hoops of setting MFA will not get `IndexError` exception?

**Comment by PageMastr** [comment]
> I did have `accessToken` and `refreshToken` in response received with my diff included, so I believe `login()` was successful.
> 
> Do you think if I will go through the hoops of setting MFA will not get `IndexError` exception?

It might be worth a shot.  the only way to know is to try, However.

Your login response, is similar to my succesful login response. So i believe your account was succesfully logging in. However, due to webull not being available in your country, your account lacks a secondary account id. which is what gives the IndexError exception.  So I do not find it likely to work. but might be worth trying.

**Comment by kucharskim** [comment]
I asked, because I do think my diff should be merged. I believe MFA or non-MFA login, code will crash with IndexError.

**Comment by PageMastr** [comment]
> I asked, because I do think my diff should be merged. I believe MFA or non-MFA login, code will crash with IndexError.

I addressed it directly on the pull, what needs to be changed for it to be accepted, to reitterate why here,

The submitted pull, will no longer cause an exception if a second account id does not exist. this would allow the users program to continue to run, making api calls.  which can cause an account lockout.

"I believe MFA or non-MFA login, code will crash with IndexError."
This is Only true in a Use case like yours where you dont have a second account id, because your not eligible to trade on webull as its not available in your country.

the majority of people who experience this error, DO have second account ids, and as such, can only get the index error, due to a failed login. Your code change, would result in those majority of users, potentially locking out their accounts after one failed login attempt. if they make a bunch of api calls or refresh calls after logging in.
